python/lsst/ctrl/oods/gen3ButlerIngester.py

```diff
@@ -40,20 +56,77 @@ def __init__(self, butlerConfig):
 
         instr = getInstrument(instrument)
         run = instr.makeDefaultRawIngestRunName()
-        opts = dict(run=run, writeable=True)
-        self.btl = Butler(butlerConfig, **opts)
+        opts = dict(run=run, writeable=True, collections=self.collections)
+        self.butler = Butler(butlerConfig, **opts)
 
         if register:
             # Register an instrument.
-            instr.register(self.btl.registry)
+            instr.register(self.butler.registry)
 
     def ingest(self, filename):
+        """Ingest a file into a butler
+
+        Parameters
+        ----------
+        filename: `str`
+            file name of the file to ingest
+        """
 
         # Ingest image.
         cfg = RawIngestConfig()
         cfg.transfer = "direct"
-        task = RawIngestTask(config=cfg, butler=self.btl)
+        task = RawIngestTask(config=cfg, butler=self.butler)
         task.run([filename])
 
     def getName(self):
+        """Return the name of this ingester
+
+        Returns
+        -------
+        ret: `str`
+            name of this ingester
+        """
         return "gen3"
+
+    async def run_task(self):
+        """run the clean() method at the configured interval
+        """
+        seconds = TimeInterval.calculateTotalSeconds(self.scanInterval)
+        while True:
+            self.clean()
+            await asyncio.sleep(seconds)
+
+    def clean(self):
+        """Remove all the datasets in the butler that
+        were ingested before the configured Interval
+        """
+
+        # calculate the time value which is Time.now - the
+        # "olderThan" configuration
+        t = Time.now()
+        interval = collections.namedtuple("Interval", self.olderThan.keys())(*self.olderThan.values())
+        td = TimeDelta(interval.days*u.d + interval.hours * u.h +
+                       interval.minutes*u.min + interval.seconds*u.s)
+        t = t - td
+
+        # get the datasets
+        ref = self.butler.registry.queryDatasets(datasetType=...,
+                                                 collections=self.collections,
+                                                 where=f"ingest_date < T'{t}'")
+
+        # retrieve the URI,  prune the dataset from
+        # the Butler, and if the URI was available,
+        # remove it.
+        for x in ref:
+            print(f"removing {x}")
+
+            uri = None
+            try:
+                uri = self.butler.getURI(x, collections=x.run)
```

Comment:
I was wondering if we could change `getURI` to return `None` if the file doesn't exist rather than raise (as an option). I thought about using `predict=True` and then checking `uri.fragment` but that would slow things down because of all the template expansion going on.

Reply:
I only have one use case for this, and this is fine for me.  Any way you think is best and makes the most sense for the use cases you know about is probably best.